internal/app/upstream/client_test.go

```diff
@@ -0,0 +1,225 @@
+package upstream_test
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"testing"
+	"time"
+
+	v2 "github.com/envoyproxy/go-control-plane/envoy/api/v2"
+	core "github.com/envoyproxy/go-control-plane/envoy/api/v2/core"
+	"github.com/envoyproxy/xds-relay/internal/app/upstream"
+	mock "github.com/envoyproxy/xds-relay/test/mocks/upstream"
+	"github.com/stretchr/testify/assert"
+)
+
+type CallOptions = upstream.CallOptions
+
+func TestOpenStreamShouldReturnErrorForInvalidTypeUrl(t *testing.T) {
+	responseChan := make(chan *v2.DiscoveryResponse)
+	client := mock.NewClient(
+		context.Background(),
+		CallOptions{Timeout: time.Nanosecond},
+		nil,
+		responseChan,
+		func(m interface{}) error { return nil })
+
+	respCh, _, err := client.OpenStream(nil)
+	assert.NotNil(t, err)
+	assert.Nil(t, respCh)
+
+	respCh, _, err = client.OpenStream(&v2.DiscoveryRequest{})
+	assert.NotNil(t, err)
+	_, ok := err.(*upstream.UnsupportedResourceError)
+	assert.True(t, ok)
+	assert.Nil(t, respCh)
+}
+
+func TestOpenStreamShouldResturnErrorOnStreamCreationFailure(t *testing.T) {
+	responseChan := make(chan *v2.DiscoveryResponse)
+	client := mock.NewClient(
+		context.Background(),
+		CallOptions{Timeout: time.Nanosecond},
+		fmt.Errorf("error"),
+		responseChan,
+		func(m interface{}) error { return nil })
+
+	typeURLs := []string{
+		upstream.ListenerTypeURL,
+		upstream.ClusterTypeURL,
+		upstream.RouteTypeURL,
+		upstream.EndpointTypeURL,
+	}
+	for _, typeURL := range typeURLs {
+		t.Run(typeURL, func(t *testing.T) {
+			respCh, _, err := client.OpenStream(&v2.DiscoveryRequest{
+				TypeUrl: typeURL,
+				Node:    &core.Node{},
+			})
+			assert.Nil(t, respCh)
+			assert.NotNil(t, err)
+		})
+	}
+}
+
+func TestOpenStreamShouldReturnNonEmptyResponseChannel(t *testing.T) {
+	responseChan := make(chan *v2.DiscoveryResponse)
+	client := mock.NewClient(
+		context.Background(),
+		CallOptions{Timeout: time.Nanosecond},
+		nil,
+		responseChan,
+		func(m interface{}) error { return nil })
+
+	respCh, done, err := client.OpenStream(&v2.DiscoveryRequest{
+		TypeUrl: upstream.ListenerTypeURL,
+		Node:    &core.Node{},
+	})
+	assert.NotNil(t, respCh)
+	assert.Nil(t, err)
+	close(done)
```

Comment:
one oddity of the ownership of done channel is that the channel is created in the xdsclient, but expects to be cleaned up in the caller.  We should solve it by changing the ownership of the channel towards the caller, but we can do it in a later iteration once we see value in it.